php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+        }
+
+        try {
+            $data = DB::arrayTable(
+                'SkillID',
+                'SELECT '.
+                '    DemonstrationSkill.*, '.
+                '    Demonstration.Demonstrated AS DemonstrationDate '.
+                '  FROM `%s` DemonstrationSkill '.
+                '  JOIN (SELECT ID, Demonstrated FROM `%s` WHERE StudentID = %u) Demonstration '.
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate, DemonstrationID',
+                [
+                    DemonstrationSkill::$tableName,
+                    Demonstration::$tableName,
+                    $this->StudentID,
+                    join(', ', $this->Competency->getSkillIds()),
+                    $this->Level
+                ]
+            );
+
+            return $this->demonstrationData = $data;
+        } catch (TableNotFoundException $e) {
+            return [];
+        }
+    }
+
+    protected static function sortEffectiveDemonstrations($a, $b)
+    {
+        if ($a['DemonstratedLevel'] == $b['DemonstratedLevel']) {
+            return 0;
+        }
+        return ($a['DemonstratedLevel'] < $b['DemonstratedLevel']) ? -1 : 1;
+    }
+
+    private $effectiveDemonstrationsData;
+    public function getEffectiveDemonstrationsData()
+    {
+        if ($this->effectiveDemonstrationsData !== null) {
+            return $this->effectiveDemonstrationsData;
+        }
+
+        $demonstrationsData = $this->getDemonstrationData();
+
+        foreach ($demonstrationsData as $skillId => &$demonstrationData) {
+            uasort($demonstrationData, [__CLASS__,  'sortEffectiveDemonstrations']);
+
+            $Skill = Skill::getByID($skillId);
+            $demonstrationsRequired = $Skill->getDemonstrationsRequiredByLevel($this->Level);
+
+            array_splice($demonstrationData, $demonstrationsRequired);
+        }
+
+        return $this->effectiveDemonstrationsData = $demonstrationsData;
+    }
+
+    private $demonstrationsLogged;
+    public function getDemonstrationsLogged()
+    {
+        if ($this->demonstrationsLogged !== null) {
+            return $this->demonstrationsLogged;
+        }
+
+        $effectiveDemonstrationsData = $this->getEffectiveDemonstrationsData();
+        $count = 0;
+        foreach ($effectiveDemonstrationsData as $skillId => $demonstrationData) {
+            $Skill = Skill::getByID($skillId);
+            $skillCount = 0;
+            foreach ($demonstrationData as $demonstration) {
+                if (empty($demonstration['Override']) && !empty($demonstration['DemonstratedLevel'])) {
+                    $skillCount++;
+                }
+            }
+            $count += $skillCount;
+        }
+
+        return $this->demonstrationsLogged = $count;
```

Comment:
I mentioned this in the last round, there's no need to use this intermediate variables. Let's work on the variable we're trying to set directly as much as possible to reduce the cognitive load of reading the code

```php

if ($this->demonstrationsLogged === null) {
    $this->demonstrationsLogged = 0;
    foreach (...) {
        $this->demonstrationsLogged += $whatever;
    }
}

return $this->demonstrationsLogged;
```